rest/blip_api_test.go

```diff
@@ -2485,7 +2485,258 @@ func TestBlipDeltaSyncPullRevCache(t *testing.T) {
+	})
+	require.NoError(t, err)
+	defer client1.Close()
+
+	// Create client with a user who have access to both channel1 and channel2
+	client2, err := NewBlipTesterClientOptsWithRT(t, rt, &BlipTesterClientOpts{
+		Username:     "bob",
+		Channels:     []string{"channel1", "channel2"},
+		ClientDeltas: true,
+	})
+	require.NoError(t, err)
+	defer client2.Close()
+
+	// Create the first revision of doc1 on both channel1 and channel2
+	resp := rt.SendAdminRequest(http.MethodPut, "/db/doc1",
+		`{"channels":["channel1","channel2"],"keys":["key1"]}`)
+	require.Equal(t, http.StatusCreated, resp.Code)
+	waitForChangesInChannel([]string{"channel1", "channel2"}, 1)
+
+	// Start one-shot pull at client1 and retrieve the first revision.
+	err = client1.StartOneshotPull()
+	assert.NoError(t, err)
+
+	data, ok := client1.WaitForRev("doc1", "1-ebebfcdd41fb6253d9ab955fa32a9791")
+	require.True(t, ok)
+	assert.Contains(t, string(data), `"channels":["channel1","channel2"]`)
+	assert.Contains(t, string(data), `"keys":["key1"]`)
+
+	// Start one-shot pull at client2 and retrieve the first revision.
+	err = client2.StartOneshotPull()
+	assert.NoError(t, err)
+
+	data, ok = client2.WaitForRev("doc1", "1-ebebfcdd41fb6253d9ab955fa32a9791")
+	require.True(t, ok)
+	assert.Contains(t, string(data), `"channels":["channel1","channel2"]`)
+	assert.Contains(t, string(data), `"keys":["key1"]`)
+
+	// Create the second revision of doc1 on channel2 as removal from channel1.
+	resp = rt.SendAdminRequest(http.MethodPut, "/db/doc1?rev=1-ebebfcdd41fb6253d9ab955fa32a9791",
+		`{"channels":["channel2"],"keys":["key1","key2"]}`)
+	assert.Equal(t, http.StatusCreated, resp.Code)
+	waitForChangesInChannel([]string{"channel2"}, 1)
+
+	// Start one-shot pull at client1 and retrieve the second revision.
+	err = client1.StartOneshotPull()
+	assert.NoError(t, err)
+
+	data, ok = client1.WaitForRev("doc1", "2-e5ee743e42d9780e22bd7b26dd4a4e60")
+	require.True(t, ok)
+	assert.Contains(t, string(data), `{"_removed":true}`)
+
+	// Create the third revision of doc1 on channel2 as leaf revision.
+	resp = rt.SendAdminRequest(http.MethodPut, "/db/doc1?rev=2-e5ee743e42d9780e22bd7b26dd4a4e60",
+		`{"channels":["channel2"],"keys":["key1","key2","key3"]}`)
+	assert.Equal(t, http.StatusCreated, resp.Code)
+	waitForChangesInChannel([]string{"channel2"}, 1)
+
+	// Get a reference to the database.
+	dbContext, err := rt.ServerContext().GetDatabase("db")
+	require.NoError(t, err, "Error getting database context")
+	database, err := db.GetDatabase(dbContext, nil)
+	require.NoError(t, err, "Error getting database")
+
+	// Flush the revision cache and purge the old revision backup.
+	database.FlushRevisionCacheForTest()
+	err = database.PurgeOldRevisionJSON("doc1", "2-e5ee743e42d9780e22bd7b26dd4a4e60")
+	require.NoError(t, err, "Error purging old revision JSON")
+
+	// Start one-shot pull at client2 and check for delta sync error.
+	err = client2.StartOneshotPull()
+	assert.NoError(t, err)
+
```

Comment:
Client 2 has access to rev 3, so I'd expect a one-shot pull replication to get rev 3 successfully.

Here I think we want to be running as Client 1.  The last revision they have access to is rev 2 (as a removal), so this would be trying to generate a delta from rev 1 to rev 2.

Reply:
Make sense, this test has been replaced with both TestDeltaSyncWhenFromRevIsChannelRemoval and TestDeltaSyncWhenToRevIsChannelRemoval to check the delta sync behavior.